src/knx/association_table_object.cpp

```diff
@@ -39,46 +39,94 @@ uint16_t AssociationTableObject::getASAP(uint16_t idx)
     return ntohs(_tableData[2 * idx + 2]);
 }
 
+// after any table change the table is checked if it allows
+// binary search access. If not, sortedEntryCount stays 0, 
+// otherwise sortedEntryCount represents size of bin search array 
+void AssociationTableObject::prepareBinarySearch()
+{
+    sortedEntryCount = 0;
+#ifdef USE_BINSEARCH
+    uint16_t lastASAP = 0;
+    uint16_t currentASAP = 0;
+    uint16_t lookupIdx = 0;
+    uint16_t lookupASAP = 0;
+    // we iterate through all ASAP
+    // the first n ASAP are sorted (strictly increasing number), these are assigning sending TSAP
+    // the remaining ASAP have to be all repetitions, otherwise we set sortedEntryCount to 0, which forces linear search 
+    for (uint16_t idx = 0; idx < entryCount(); idx++)
+    {
+        currentASAP = getASAP(idx);
+        if (sortedEntryCount)
+        {
+            // look if the remaining ASAP exist in the previously sorted list.
+            while (lookupIdx < sortedEntryCount)
+            {
+                lookupASAP = getASAP(lookupIdx);
+                if (currentASAP <= lookupASAP)
+                    break; // while
+                else
+                    lookupIdx++;
+            }
+            if (currentASAP < lookupASAP || lookupIdx >= sortedEntryCount)
+            {
+                // a new ASAP found, we force linear search
+                sortedEntryCount = 0;
+                break; // for
+            }
+        }
+        else
+        {
+            // check for strictly increasing ASAP
+            if (currentASAP > lastASAP)
+                lastASAP = currentASAP;
+            else
+            {
+                sortedEntryCount = idx; // last found index indicates end of sorted list
+                idx--; // current item has to be handled as remaining ASAP
+            }
+        }
+    }
+    // in case complete table is strictly increasing
+    if (lookupIdx == 0 && sortedEntryCount == 0)
+        sortedEntryCount = entryCount();
+#endif    
+}
+
 const uint8_t* AssociationTableObject::restore(const uint8_t* buffer)
 {
     buffer = TableObject::restore(buffer);
     _tableData = (uint16_t*)data();
+    prepareBinarySearch();
     return buffer;
 }
 
 // return type is int32 so that we can return uint16 and -1
 int32_t AssociationTableObject::translateAsap(uint16_t asap)
 {
     uint16_t entries = entryCount();
-    #ifdef USE_BINSEARCH
-    uint16_t low,high,i;
-    low = 0;
-    high = entries-1;
-
-    while(low <= high)
+    if (sortedEntryCount)
```

Comment:
I think you should add a comment here that you rely on strong monotone sorting here, and we know it is because of the prepareBinSearch method.

Reply:
Hi, I added the required comments. I hope they explain my approach.

Regards,
Waldemar